### PR TITLE
MINOR: Log transaction metadata state transitions plus a few cleanups

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

File: core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala [hunk 1]
```diff
@@ -98,13 +104,24 @@ private[transaction] object TransactionMetadata {
...
-                                                              txnStartTimestamp: Long,
-                                                              txnLastUpdateTimestamp: Long)
+private[transaction] case class TxnTransitMetadata(producerId: Long,
+                                                   producerEpoch: Short,
+                                                   txnTimeoutMs: Int,
+                                                   txnState: TransactionState,
+                                                   topicPartitions: immutable.Set[TopicPartition],
+                                                   txnStartTimestamp: Long,
+                                                   txnLastUpdateTimestamp: Long) {
+  override def toString: String = {
```
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:114]: Do we need to override case class' `toString`? I think they are auto-provided already?

File: core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala [hunk 2]
```diff
@@ -233,70 +245,83 @@ private[transaction] class TransactionMetadata(val producerId: Long,
...
+      s"state=$state, " +
+      s"pendingState=$pendingState, " +
+      s"topicPartitions=$topicPartitions, " +
+      s"txnStartTimestamp=$txnStartTimestamp, " +
+      s"txnLastUpdateTimestamp=$txnLastUpdateTimestamp)"
+  }
 
   override def equals(that: Any): Boolean = that match {
     case other: TransactionMetadata =>
+      transactionalId == other.transactionalId &&
```
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:324]: This field is added just for logging? In that case we could just add the logging in the caller of `completeTransitionTo` and `prepareTransitionTo`.

File: core/src/test/scala/unit/kafka/coordinator/transaction/TransactionCoordinatorTest.scala [hunk 3]
```diff
@@ -642,7 +666,7 @@ class TransactionCoordinatorTest {
     EasyMock.expect(transactionManager.appendTransactionToLog(
       EasyMock.eq(transactionalId),
       EasyMock.eq(coordinatorEpoch),
-      EasyMock.eq(originalMetadata.copy().prepareAbortOrCommit(transactionState, now)),
+      EasyMock.eq(transition),
```
Comment (guozhangwang) [inline core/src/test/scala/unit/kafka/coordinator/transaction/TransactionCoordinatorTest.scala:669]: Nice catch!

File: core/src/test/scala/unit/kafka/coordinator/transaction/TransactionStateManagerTest.scala [hunk 4]
```diff
@@ -62,21 +62,23 @@ class TransactionStateManagerTest {
...
-  val txnMessageKeyBytes2: Array[Byte] = TransactionLog.keyToBytes(txnId2)
-  val pidMappings: Map[String, Long] = Map[String, Long](txnId1 -> 1L, txnId2 -> 2L)
-  var txnMetadata1: TransactionMetadata = TransactionMetadata(pidMappings(txnId1), 1, transactionTimeoutMs, 0)
-  var txnMetadata2: TransactionMetadata = TransactionMetadata(pidMappings(txnId2), 1, transactionTimeoutMs, 0)
+  val transactionalId1: String = "one"
+  val transactionalId2: String = "two"
+  val txnMessageKeyBytes1: Array[Byte] = TransactionLog.keyToBytes(transactionalId1)
+  val txnMessageKeyBytes2: Array[Byte] = TransactionLog.keyToBytes(transactionalId2)
+  val pidMappings: Map[String, Long] = Map[String, Long](transactionalId1 -> 1L, transactionalId2 -> 2L)
+  var txnMetadata1: TransactionMetadata = TransactionMetadata(transactionalId1, pidMappings(transactionalId1), 1,
```
Comment (guozhangwang) [inline core/src/test/scala/unit/kafka/coordinator/transaction/TransactionStateManagerTest.scala]: nit: We can use newly added `transactionMetadata` here as well.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala (hunk 1 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:114]: The problem is that we don't get the field names in the default toString.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala (hunk 2 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:324]: We could, but that's kind of annoying, right? And it's nice to ensure we have all cases covered. It's sort of like how we ultimately added `groupId` to `GroupMetadata`.

**Comment by guozhangwang** [comment]
LGTM. Please rebase :)

**Pull request closed without merging**